CmdStarter.Lib/Attributes/AutoComplete/AutoCompleteAttribute.cs

```diff
@@ -0,0 +1,164 @@
+﻿using com.cyberinternauts.csharp.CmdStarter.Lib.Interfaces;
+using System.CommandLine.Completions;
+
+namespace com.cyberinternauts.csharp.CmdStarter.Lib.Attributes
+{
+    /// <summary>
+    /// Defines auto-completions for an <see cref="System.CommandLine.Option"/> or an <see cref="System.CommandLine.Argument"/>.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Parameter, AllowMultiple = true)]
+    public class AutoCompleteAttribute : Attribute
+    {
+        private const string NULL_OR_EMPTY_ERROR_MESSAGE = "A completion cannot be null or empty!";
+
+        private readonly string[] _labels;
+        private readonly IAutoCompleteFactory? _factory;
+        private LinkedList<CompletionItem>? _items;
+
+        public CompletionDelegate Context
+        {
+            get
+            {
+                if (_items is null) CacheItems();
+
+                return (ctx) => _items!; //Not null as CacheItems() initializes it.
+            }
+        }
+
+        /// <summary>
+        /// Generates completion from a given provider.
+        /// <para>
+        /// If <paramref name="provider"/> is <see langword="typeof"/> <see cref="Enum"/>
+        /// generates auto completions from all values. 
+        /// </para>
+        /// <para>
+        /// If <paramref name="provider"/> is <see langword="typeof"/> <see cref="IAutoCompleteProvider"/>
+        /// retrieves auto completions from there.
+        /// </para>
+        /// </summary>
+        /// <param name="provider"><see cref="Type"/> of an <see cref="Enum"/> or an <see cref="IAutoCompleteProvider"/>.</param>
+        public AutoCompleteAttribute(Type provider)
+            : this(HandleProviderType(provider))
+        { }
+
+        /// <summary>
+        /// Generates completion from a given provider and runs them through the <paramref name="factory"/>.
+        /// <para>
+        /// If <paramref name="provider"/> is <see langword="typeof"/> <see cref="Enum"/>
+        /// generates auto completions from all values. 
+        /// </para>
+        /// <para>
+        /// If <paramref name="provider"/> is <see langword="typeof"/> <see cref="IAutoCompleteProvider"/>
+        /// retrieves auto completions from there.
+        /// </para>
+        /// </summary>
+        /// <param name="factory">Must be <see langword="typeof"/> <see cref="IAutoCompleteFactory"/>.</param>
+        /// <param name="provider"><see cref="Type"/> of an <see cref="Enum"/> or an <see cref="IAutoCompleteProvider"/>.</param>
+        public AutoCompleteAttribute(Type factory, Type provider)
```

Comment:
I don't think it's that error prone.

I think the documentation gives a great context about what those parameters should be.

Reply:
Well, I agree, if the class type used only implements `IAutoCompleteFactory`, it is useless. And it could throw an exception.

But one could implement both interfaces: provider and factory.

In fact, this is if we support using `Type` (like currently we are). Otherwise, we move all the type specific code to the generic version, remove the constructors with `Type` in non-generic and use `<Provider, Factory>... where ENSURE_GENERICS`. But this solution offers less options.

Another argument in favor of removing this constructor would be: If we would like to support a more than one type that each could bring a functionality, I would implement a list instead of only two.

My idea is that minimal is better. If there were two abstract classes it would be mandatory, but no, we are using interfaces. They all can be merged in one. (Not from our side, but the user side). Still, from our side, if we find other things to implement, we may create a list of classes supporting each its feature and this type would be passed all around to execute the specific code of that feature. If we start having more than one, meaning one is for this feature, and the other for that feature. It might be harder to extend.

What do you think?

Reply:
I like your thoughts of one type implementing two interfaces but wouldn't it be limiting if someone was for example closely following SOLID?

I think we could extend the one `Type` to check for factory too.

Reply:
Here, the SOLID principle can still be achieve by the user with a class called `AutoCompleteFeaturesProxy` that would implement all the desired features by calling the appropriate classes that support it.

And, on our side, **later**, we could support a mechanism to support multiple types in a constructor and correctly managing collision cases (like two types implementing `IAutoCompleteFactory` (if a collision is really a problem).

Reply:
Okay, let's remove this constructor and give support to the one type constructor then.